# Homework 3
### DS7337 - 404

Jason Rupp

In [16]:
import numpy as np 
import pandas as pd 
from nltk.metrics import edit_distance
from nltk.corpus import words, stopwords
import re
from nltk.stem import PorterStemmer

class color:
   PURPLE = '\033[95m'
   CYAN = '\033[96m'
   DARKCYAN = '\033[36m'
   BLUE = '\033[94m'
   GREEN = '\033[92m'
   YELLOW = '\033[93m'
   RED = '\033[91m'
   BOLD = '\033[1m'
   UNDERLINE = '\033[4m'
   END = '\033[0m'
    
# https://stackoverflow.com/questions/8924173/how-do-i-print-bold-text-in-python


#### 1. Compare your given name with your nickname (if you don’t have a nickname, invent one for this assignment) by answering the following questions:

a. What is the edit distance between your nickname and your given name?    
b. What is the percentage string match between your nickname and your given name?
    
Show your work for both calculations.


In [2]:
def hb_edit_distance(str1, str2):
    '''
    function created with the help of the following resource:
    https://www.w3resource.com/python-exercises/challenges/1/python-challenges-1-exercise-52.php    
    '''
    # normalize text
    str1=str1.lower()
    str2=str2.lower()
    
    # these vars will be used to assign the longer word
    longer = ""
    shorter = ""
    
    # if the lens differ, the edit distance starts there
    if len(str1)>len(str2):
        diff = len(str1) - len(str2)
        longer = str1
        shorter = str2
    elif len(str2) > len(str1):
        diff = len(str2) - len(str1)
        longer = str2
        shorter = str1
    else:
        diff = 0    
    

    # now compare differences in char
    for i in range(len(shorter)):
        if shorter[i] != longer[i]:
            diff += 1
    
    return diff

In [70]:
# edit distance between name and nick name
given_name = "Jason"

nick_name = "Jay"

nickNameDiff = hb_edit_distance(given_name,nick_name)

print(nickNameDiff)

3


In [78]:
# find percentage string match

string_match = ((len(given_name) - nickNameDiff)/len(given_name)) * 100

print("Percent String Match: ",string_match, "%", sep = "")

40.0%


The edit distance between my given name, Jason and nick name, Jay is 3. The percentage string match can be found by subtracting the edit distance from the total length of my given name and finding the percentage.

#### 2.	Find a friend (or family member or classmate) who you know has read a certain book. Without your friend knowing, copy the first two sentences of that book. Now rewrite the words from those sentences, excluding stop words. Now tell your friend to guess which book the words are from by reading them just that list of words. Did you friend correctly guess the book on the first try? What did he or she guess? Explain why you think you friend either was or was not able to guess the book from hearing the list of words. 

*There was only orange juice in the fridge. Nothing else that you could put on cereal, unless you think that ketchup or mayonnaise or pickle juice would be nice on your Toastios, which I do not, and neither did my little sister, although she has eaten some pretty weird things in her day, like mushrooms in chocolate.*



Fortunately, the Milk

By: Neil Gaiman

In [4]:
sentence1 = "There was only orange juice in the fridge."
sentence2 = """Nothing else that you could put on cereal, 
unless you think that ketchup or mayonnaise or pickle juice would be nice on your Toastios, 
which I do not, and neither did my little sister, 
although she has eaten some pretty weird things in her day, 
like mushrooms in chocolate."""



In [9]:
def stop_drop(text, sort = False):
    tokens= re.findall('\w+', text.lower())
    
    stops = stopwords.words('english')
    stops_dropped = [i for i in tokens if i not in stops]
    
    if sort:
        stops_dropped = sorted(stops_dropped)
    
    return(stops_dropped)

In [10]:
s1_stops_dropped = stop_drop(sentence1)
s2_stops_dropped = stop_drop(sentence2)

In [34]:
print(color.BOLD + "First sentence:" + color.END, *s1_stops_dropped)

First sentence: orange juice fridge


In [35]:
print(color.BOLD + "Second sentence:" + color.END, *s2_stops_dropped)

Second sentence: nothing else could put cereal unless think ketchup mayonnaise pickle juice would nice toastios neither little sister although eaten pretty weird things day like mushrooms chocolate


***

##### Response:

The book I chose was from "Fortunately, the Milk" by Neil Gaiman. It is a children's book of about 100 pages that my daughter has read several times. She was able to recognize which book these two sentences came from within a matter of seconds, she even claimed to have known from the first sentence with just "orange juice fridge". 

The stop words don't add to the context of the sentence. The main themes are preserved after the stop words are removed, which is why she was able to guess correctly which book it came from.

#### 3.	Run one of the stemmers available in Python. Run the same two sentences from question 2 above through the stemmer and show the results. How many of the outputted stems are valid morphological roots of the corresponding words? Express this answer as a percentage.

In [43]:
ps = PorterStemmer()

first_two = s1_stops_dropped + s2_stops_dropped

ps_stemmed = [ps.stem(i) for i in first_two]

print(color.BOLD + "Stemmed Sentences:\n" + color.END, *ps_stemmed)

Stemmed Sentences:
 orang juic fridg noth els could put cereal unless think ketchup mayonnais pickl juic would nice toastio neither littl sister although eaten pretti weird thing day like mushroom chocol


In [44]:
def calc_real_word(text):
    
    # initialize dictionary 
    d = {
    'realWords': [i for i in text if i in words.words()],
    'realWord_count': sum([1 for i in text if i in words.words()]),
    'totalWord_count':len(text)       
    } 
    
    d['realPercent'] = (d['realWord_count']/d['totalWord_count']) * 100

    return(d)

In [45]:
stemmed_sentences = calc_real_word(ps_stemmed)

In [51]:
print("Percentage of stems that are words: ", round(stemmed_sentences['realPercent'],2), "%", sep = "")

Percentage of stems that are words: 65.52%


In [55]:
for i in range(len(stemmed_sentences['realWords'])):
    print(stemmed_sentences['realWords'][i])

orang
els
could
put
cereal
unless
think
ketchup
would
nice
neither
sister
although
eaten
weird
thing
day
like
mushroom


In [66]:
adjusted_percent = (stemmed_sentences['realWord_count'] - 1)/stemmed_sentences['totalWord_count'] * 100

In [67]:
print("Adjusted percentage of stems that are words: ", round(adjusted_percent,2), "%", sep = "")

Percentage of stems that are words: 62.07%


The above percentage (~65%) is being calculated by comparing the stemmed words from the sentences, and comparing them the the English language words corpus available in the nltk package. If the word is found in the corpus it is considered valid. An inspection of the list chosen does possibly raise an issue or two. The first two words on the list below were considered real words.


A quick google search reveals that orang is a <a href="https://www.google.com/search?q=orang&oq=orang&aqs=chrome.0.69i59j35i39j46i433j46i67i433j46i433l2j69i60l2.1959j0j7&sourceid=chrome&ie=UTF-8">valid word</a>, though it seems like it is has a very uncommon usage. 

The second "word" on this list is "els" and there doesn't seem to be any such evidence for this being a word, only usage found was in an acronym. So it can be eliminate from this list and recalculated.

The final percentage is about 62% of the word roots were valid words.